oeps/oep-0019-bp-developer-documentation.rst

```diff
@@ -0,0 +1,293 @@
+.. _invaluable technique: https://www.thoughtworks.com/radar/techniques/lightweight-architecture-decision-records
+.. _local decisions: https://github.com/edx/credentials/tree/master/docs/decisions
+.. _Swagger: https://swagger.io/
+.. _Sphinx: http://www.sphinx-doc.org/en/master/
+
+Tools and Location
+~~~~~~~~~~~~~~~~~~
+
+The recommended tools and locations for developer documentation are described below, inlined with the
+requirements they meet.
+
+* **Discoverable**. Users must be able to find the documentation that is relevant to their needs. This 
+  calls for a high-level index and search mechanism for finding content.
+
+  * **Search**. For the time-being, we will make use of Google and Github's search functionality for
+    finding content since all developer documentation will be in Github.  Certain IDEs may also have good
+    tools for searching.
+
+  * **Index**. In the future, we will invest time in a better automated indexing tool to aggregate and
+    structure the documentation. For the time being, we will manually create and maintain documentation
+    indices at:
+    
+    * edx-developer-docs_ will include the landing page for developer documentation and be a container of
+      orphaned and cross-repo developer documentation.
+
+    * docs.edx.org_ will be the main landing page for *all* Open edX documentation, including a link to the
+      developer documentation index listed above.
+
+* **Maintainable**. Writers, editors, and reviewers must be able to create and modify documentation without
+  too much effort. Anyone, inside or outside of edX, should have the ability to contribute to documentation.
+
+  * **Co-located**. The closer the documentation lives with its corresponding code, the more likely it will be
+    kept up-to-date.
+
+    * Repo-specific documentation will live within a "docs" folder within its repo. This includes local
+      "decisions", "how-tos", and README.rst. Documentation should not live within the code itself (e.g., in
+      an __init__.py python module).
+
+    * System-wide decisions will live within the open-edx-proposals_ repo.
+
+    * Other system-wide documentation will live within the edx-developer-docs_ repo.
+
+  * **Exported Diagrams**. In order to keep diagrams updated across time, they will need to be exported
+    from whichever tool they were created in, with the exported version maintained in Github.
+
+    * draw.io diagrams will be exported and imported as XML files.
+    * Lucidchart diagrams will be exported and imported as Visio (VDX) files.
+
+* **Version Controlled**. Documentation must be maintained under version control in Github, for the
+  following reasons:
+
+  * **Reviewed**. They will be properly reviewed as defined by that repo's process.
+
+  * **Tagged**. They will be appropriately in-synch with its corresponding code. When an open edX instance
+    is running a specific named release, they can refer to the matching documentation for that version of the
```

Comment:
Does this mean we can look forward to simply navigating to the specific generated and hosted docs that match the software version of edx-platform, ecommerce, etc?

Reply:
@johnbaldwin raises a good point. Part of the frustration with the current Open edX docs, is that it's very difficult to tell which docs go with which version of edx-platform. In many cases, one realizes that the docs have been deprecated and are no longer relevant to the version of Open edX that you're using, but there is no indicator in the docs themselves that this is the case.

Maybe if there were an identifier at the top of each doc that said, "These docs correspond to the Hawthorn release of Open edX" that would help immensely.

Reply:
@johnbaldwin Yes, that's the objective.  When developer docs are stored within Git repos, they can be updated alongside the code.  This means if someone is using an older `Open edX named release`, they can reference the dev docs for that particular release - regardless of how the `master` branch has moved forward since the co-located docs should be in synch with the code.